File: app/src/main/java/ru/andrey/cleandictionary/di/translation/TranslationModule.java
package ru.andrey.cleandictionary.di.translation;

import dagger.Module;
import dagger.Provides;
import ru.andrey.cleandictionary.domain.global.TranslationRepository;
import ru.andrey.cleandictionary.domain.global.TranslationService;
import ru.andrey.cleandictionary.domain.interactor.favorite.FavoriteTranslationInteractor;
import ru.andrey.cleandictionary.domain.interactor.favorite.FavoriteTranslationInteractorImpl;
import ru.andrey.cleandictionary.domain.interactor.translation.TranslationInteractor;
import ru.andrey.cleandictionary.domain.interactor.translation.TranslationInteractorImpl;
import ru.andrey.cleandictionary.domain.interactor.translationlist.TranslationsListInteractor;
import ru.andrey.cleandictionary.domain.interactor.translationlist.TranslationsListInteractorImpl;

@Module
public class TranslationModule {

    @Provides
    @TranslationScope
    FavoriteTranslationInteractor
    provideFavoriteTranslationInteractor(TranslationInteractor interactor) {
        return new FavoriteTranslationInteractorImpl(interactor);
    }

    @Provides
    @TranslationScope
    TranslationInteractor
    provideTranslationInteractor(TranslationService service, TranslationRepository repository) {
        return new TranslationInteractorImpl(service, repository);
    }

    @Provides
    @TranslationScope
    TranslationsListInteractor provideTranslationsListInteractor(TranslationRepository repository) {
        return new TranslationsListInteractorImpl(repository);
    }
}

File: presentation/src/main/java/ru/andrey/cleandictionary/di/translation/TranslationModule.kt
package ru.andrey.cleandictionary.di.translation

import dagger.Module
import dagger.Provides
import io.reactivex.schedulers.Schedulers
import retrofit2.Retrofit
import ru.andrey.cleandictionary.di.scope.Feature
import ru.andrey.data.api.MicrosoftApi
import ru.andrey.data.api.YandexApi
import ru.andrey.data.db.TranslationDatabase
import ru.andrey.data.repository.MicrosoftTranslationService
import ru.andrey.data.repository.TranslationRepositoryImpl
import ru.andrey.data.repository.YandexTranslationService
import ru.andrey.domain.interactor.FavoriteInteractor
import ru.andrey.domain.interactor.TranslationInteractor
import ru.andrey.domain.interactor.TranslationListInteractor
import ru.andrey.domain.repository.TranslationRepository
import ru.andrey.domain.repository.TranslationService

@Module
open class TranslationModule(val useMicrosoft: Boolean) {

    @Provides
    @Feature
    fun provideTranslationRepository(database: TranslationDatabase): TranslationRepository {
        return TranslationRepositoryImpl(database)
    }

    @Provides
    @Feature
    open fun provideTranslationService(retrofit: Retrofit): TranslationService {
        if (useMicrosoft) {
            return MicrosoftTranslationService(retrofit.create(MicrosoftApi::class.java))
        }
        return YandexTranslationService(retrofit.create(YandexApi::class.java))
    }

    @Feature
    @Provides
    fun provideTranslationInteractor(service: TranslationService): TranslationInteractor {
        return TranslationInteractor(Schedulers.io(), service)
    }

    @Feature
    @Provides
    fun provideTranslationsListInteractor(repo: TranslationRepository): TranslationListInteractor {
        return TranslationListInteractor(Schedulers.io(), repo)
    }

    @Feature
    @Provides
    fun provideFavoriteInteractor(repo: TranslationRepository): FavoriteInteractor {
        return FavoriteInteractor(Schedulers.io(), repo)
    }
}
